### PR TITLE
Bug 1169379 - Use favicons for top sites tiles

### PR DESCRIPTION
**Pull request by wesj**

### CODE REVIEW

File: Client/Frontend/Home/TopSitesPanel.swift [hunk 1]
```diff
@@ -259,11 +264,11 @@ private class TopSitesLayout: UICollectionViewLayout {
 
     private func setupForOrientation(orientation: UIInterfaceOrientation) {
         if orientation.isLandscape {
```
Comment (sleroux) [inline Client/Frontend/Home/TopSitesPanel.swift]: I wonder if it's possible to achieve the same logic without using the orientation and only the size classes. Now that we have multi-window mode on iPad I'd like to stray away from relying on orientation if we can. Not sure if it's possible though.

File: Client/Frontend/Home/TopSitesPanel.swift [hunk 2]
```diff
@@ -348,24 +354,57 @@ private class TopSitesDataSource: NSObject, UICollectionViewDataSource {
...
     }
 
+    private func getFavicon(cell: ThumbnailCell, site: Site) {
+        // TODO: This won't work well with recycled views. Thankfully, TopSites doesn't really recycle much.'
+        cell.imageView.image = nil
+        cell.backgroundImage.image = nil
+
+        FaviconFetcher.getForUrl(site.url.asURL!, profile: profile) { icons in
+            if (icons.count > 0) {
+                cell.imageView.sd_setImageWithURL(icons[0].url.asURL!) { (img, err, type, url) -> Void in
```
Comment (sleroux) [inline Client/Frontend/Home/TopSitesPanel.swift]: Nit: Since we're downloading this url information, could it be possible that it's an invalid url and the force asURL! would crash? Just want to make sure we don't crash if someone feeds us a garbage url.

File: Client/Frontend/Widgets/ThumbnailCell.swift [hunk 3]
```diff
@@ -36,47 +46,101 @@ class ThumbnailCell: UICollectionViewCell {
...
+    lazy var imageWrapper: UIView = {
+        let imageWrapper = UIView()
         imageWrapper.layer.borderColor = ThumbnailCellUX.BorderColor.CGColor
-        imageWrapper.layer.borderWidth = 1
-        imageWrapper.layer.cornerRadius = 3
+        imageWrapper.layer.borderWidth = ThumbnailCellUX.BorderWidth
+        imageWrapper.layer.cornerRadius = ThumbnailCellUX.CornerRadius
         imageWrapper.clipsToBounds = true
+        return imageWrapper
+    } ()
```
Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: Nit: Remove space between } ()

File: Client/Frontend/Widgets/ThumbnailCell.swift [hunk 4]
```diff
@@ -36,47 +46,101 @@ class ThumbnailCell: UICollectionViewCell {
...
 
+    lazy var removeButton: UIButton = {
+        let removeButton = UIButton()
         removeButton.setImage(UIImage(named: "TileCloseButton"), forState: UIControlState.Normal)
         removeButton.addTarget(self, action: "SELdidRemove", forControlEvents: UIControlEvents.TouchUpInside)
         removeButton.hidden = true
         removeButton.imageEdgeInsets = ThumbnailCellUX.RemoveButtonInsets
+        return removeButton
+    } ()
```
Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: Nit: Remove space between } ()

File: Client/Frontend/Widgets/ThumbnailCell.swift [hunk 5]
```diff
@@ -36,47 +46,101 @@ class ThumbnailCell: UICollectionViewCell {
...
+
+    override init(frame: CGRect) {
+        imagePadding = CGFloat(0)
```
Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: Since we're setting up the constraints below, we probably don't want the didSet handler to get called whenever we set this to 0 here. Instead we set it to 0 in the declaration. Also, I don't see it in this diff but inside the didSet handler, after we call snp_remakeConstraints on the imageView, we should setNeedsUpdateConstraints on the imageWrapper to notify it of the constraint change on it's next layout pass.

File: Client/Frontend/Widgets/ThumbnailCell.swift [hunk 6]
```diff
@@ -36,47 +46,101 @@ class ThumbnailCell: UICollectionViewCell {
...
+        })
+
+        imageView.snp_remakeConstraints({ make in
+            let imagePadding: CGFloat = (UIScreen.mainScreen().traitCollection.horizontalSizeClass == .Compact) ? ThumbnailCellUX.ImagePaddingCompact : ThumbnailCellUX.ImagePadding
+            let insets = UIEdgeInsetsMake(imagePadding, imagePadding, imagePadding, imagePadding)
+            make.top.left.right.equalTo(self.imageWrapper).insets(insets)
+            make.bottom.equalTo(textWrapper.snp_top).offset(-imagePadding) // .insets(insets)
+        })
+
+        textWrapper.snp_remakeConstraints({ make in
```
Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: Nit: Instead of using snp_remakeConstraints, you can call snp_makeConstraints since you're not remaking them at this point

File: Client/Frontend/Widgets/ThumbnailCell.swift [hunk 7]
```diff
@@ -130,7 +197,9 @@ class ThumbnailCell: UICollectionViewCell {
     var imagePadding: CGFloat {
         didSet {
             imageView.snp_remakeConstraints({ make in
-                make.top.bottom.left.right.equalTo(self.imageWrapper).insets(UIEdgeInsetsMake(imagePadding, imagePadding, imagePadding, imagePadding))
+                let insets = UIEdgeInsetsMake(imagePadding, imagePadding, imagePadding, imagePadding)
+                make.top.left.right.equalTo(self.imageWrapper).insets(insets)
+                make.bottom.equalTo(textWrapper.snp_top).offset(-imagePadding) // .insets(insets)
             })
```
Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: After we call snp_remakeConstraints on the imageView, we should setNeedsUpdateConstraints on the imageWrapper to notify it of the constraint change on it's next layout pass.

Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: Also super Nit: but could you move the var declarations up with the other var declarations?

File: Client/Frontend/Widgets/ThumbnailCell.swift [hunk 8]
```diff
@@ -139,11 +208,18 @@ class ThumbnailCell: UICollectionViewCell {
         didSet {
             if let image = image {
                 imageView.image = image
-                imageView.alignment = UIImageViewAlignmentMaskTop
-                imageView.contentMode = UIViewContentMode.ScaleAspectFill
+                imageView.contentMode = UIViewContentMode.ScaleAspectFit
+
+                // Force nearest neighbor scaling for small favicons
+                if image.size.width < 24 {
```
Comment (sleroux) [inline Client/Frontend/Widgets/ThumbnailCell.swift]: What's the significance of 24? Could we move this into the UX struct?

File: Storage/Logins.swift [hunk 9]
```diff
@@ -139,7 +139,7 @@ public class Login: Printable, LoginData, LoginUsageData, Equatable {
             let url1 = NSURL(string: value!)
 
             if url1?.host != url2?.host {
-                assertionFailure("Form submit URL domain doesn't match login's domain.")
+                // assertionFailure("Form submit URL domain doesn't match login's domain.")
```
Comment (sleroux) [inline Storage/Logins.swift]: Do we need this as part of this patch?

File: Storage/SQL/BrowserDB.swift [hunk 10]
```diff
@@ -57,7 +57,7 @@ enum TableResult {
 public class BrowserDB {
     private var db: SwiftData
     // XXX: Increasing this should blow away old history, since we currently don't support any upgrades.
-    private let Version: Int = 7
+    private let Version: Int = 8
```
Comment (sleroux) [inline Storage/SQL/BrowserDB.swift]: I wonder if we should add a Github label to indicate that this PR will cause this effect?

File: Storage/SQL/BrowserTable.swift [hunk 11]
```diff
@@ -55,16 +55,109 @@ private let log = XCGLogger.defaultInstance()
...
+            // There's no point tracking rows that are not flagged for upload.
+            return "CREATE INDEX IF NOT EXISTS \(IndexHistoryShouldUpload) " +
+            "ON \(TableHistory) (should_upload) WHERE should_upload = 1"
+        } else {
+            return "CREATE INDEX IF NOT EXISTS \(IndexHistoryShouldUpload) " +
+            "ON \(TableHistory) (should_upload)"
+        }
+    }
+
+
```
Comment (sleroux) [inline Storage/SQL/BrowserTable.swift]: This is just moving code right?

File: Storage/SQL/BrowserTable.swift [hunk 12]
```diff
@@ -240,17 +243,46 @@ public class BrowserTable: Table {
...
-            "title TEXT" +
-            ") "
-            if !self.run(db, queries: [queue]) {
+            if !self.run(db, queries: [CreateQueueTable]) {
+                return false
+            }
+        }
+
+        // Moved the date column from the favicons table to the faviconSites table.
+        if (from < 6) {
```
Comment (sleroux) [inline Storage/SQL/BrowserTable.swift]: Just curious, is this how we're handling migrations right now?

File: Utils/FaviconFetcher.swift [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    // Loads favicon.ico on the host domain for this url
+    private func loadFromHost() {
+        var url = NSURL(scheme: siteUrl.scheme!, host: siteUrl.host, path: "/favicon.ico")
+        var request = NSMutableURLRequest(URL: url!)
+        request.HTTPMethod = "HEAD"
+        request.timeoutInterval = 5
+
+        var response: NSURLResponse? = nil;
+        var err = NSErrorPointer()
+        NSURLConnection.sendSynchronousRequest(request, returningResponse: &response, error: err)
```
Comment (sleroux) [inline Utils/FaviconFetcher.swift]: This method and I think most of NSURLConnection is actually deprecated starting in iOS 9 so we'll need to use NRURLSesson/Alamofire for this instead.

File: Utils/FaviconFetcher.swift [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        if let data = NSData(contentsOfURL: siteUrl),
+           let element = RXMLElement(fromHTMLData: data) {
+            element.iterate("head.meta") { meta in
+                if let refresh = meta.attribute("http-equiv"),
+                   let content = meta.attribute("content"),
+                   let index = content.rangeOfString("URL="),
+                   let url = NSURL(string: content.substringFromIndex(advance(index.startIndex,4))) {
+                    if refresh == "Refresh" {
+                        self.siteUrl = url
+                        return self.loadFromDoc()
```
Comment (sleroux) [inline Utils/FaviconFetcher.swift]: Nit: This method returns Void but we're calling the method from the return statement. Can we move this up just to make it more clear?

File: Utils/FaviconFetcher.swift [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+                fav.width = Int(img.size.width)
+                fav.height = Int(img.size.height)
+                profile.favicons.addFavicon(fav, forSite: site)
+                callback(fav)
+                return
+            }
+
+            fav.width = 0
+            fav.height = 0
+            return callback(fav)
```
Comment (sleroux) [inline Utils/FaviconFetcher.swift]: See above nit about returning the callback

File: Storage/Logins.swift (hunk 9 above)
Comment (wesj) [inline Storage/Logins.swift]: Heh. No. I have some data in my syncing profile that tickles this and causes crashes. I'll remove it.

File: Storage/SQL/BrowserTable.swift (hunk 11 above)
Comment (wesj) [inline Storage/SQL/BrowserTable.swift]: Yeah. I moved all this to a separate PR since its not really required here.

File: Storage/SQL/BrowserTable.swift (hunk 12 above)
Comment (wesj) [inline Storage/SQL/BrowserTable.swift]: I'd like to make something a little nicer, but I'm fine getting there incrementally.

File: Client/Frontend/Home/TopSitesPanel.swift (hunk 1 above)
Comment (wesj) [inline Client/Frontend/Home/TopSitesPanel.swift]: I... don't think it is. Basically we want to show 4 in portrait on tablets and 5 in landscape, but neither the horizontal (or vertical) size class changes there. I did switch the order of the checks here so that regardless of orientation, if a tablet moves into horizontal-size-class == .Compact, we'll switch to showing 3.

File: Storage/SQL/BrowserDB.swift (hunk 10 above)
Comment (wesj) [inline Storage/SQL/BrowserDB.swift]: I'm going to remove this entire commit from here. Its better in a separate PR I think.
